src/visitor.rs

```diff
@@ -0,0 +1,769 @@
+/// impl TypeResolver for MyTypeResolver {
+///     type TypeId = u32;
+///     type Error = u8;
+///     fn resolve_type<'this, V: ResolvedTypeVisitor<'this, TypeId = Self::TypeId>>(
+///         &'this self,
+///         type_id: &Self::TypeId,
+///         visitor: V,
+///     ) -> Result<V::Value, Self::Error> {
+///         Ok(visitor.visit_not_found())
+///     }
+/// }
+///
+/// // Now, we can create a visitor using this `visitor::new` function.
+/// // This has specific handling for composite and variant types, falling
+/// // back to returning `1u64` if some other type was found.
+/// let context = ();
+/// let visitor = scale_type_resolver::visitor::new(context, |_context, _unhandled_kind| 1u64)
+///     .visit_composite(|_context, _composite_fields| 2)
+///     .visit_primitive(|_context, _primitive_type| 3);
+///
+/// // By providing the visitor to a type resolver, the TypeId type can be
+/// // inferred (else it can be given on the `visitor::new` function).
+/// MyTypeResolver.resolve_type(&123, visitor);
+/// ```
+///
+/// The `visit_*` methods provided each take closures which have a similar type signature to the
+/// underlying trait methods on [`ResolvedTypeVisitor`], with small differences where necessary to
+/// avoid type and ownership issues. The first argument to every function is some arbitrary context
+/// which is provided as the first argument to [`crate::visitor::new()`].
+///
+/// Using this concrete visitor is expected to be almost as optimal as implementing the
+/// [`ResolvedTypeVisitor`] trait manually. One area where it makes a small sacrifice to this is in
+/// the [`ConcreteResolvedTypeVisitor::visit_variant()`] method, which must collect the variant
+/// fields into a [`smallvec::SmallVec`] to avoid ownership issues. This will only allocate if more
+/// than 16 variant fields exist. The default "unhandled" function here must also implement `Clone`,
+/// which isn't necessary in a raw trait implementation, since it will, by default, be called in all
+/// of the other impls if no methods are provided.
+pub fn new<'resolver, Context, TypeId, Output, NewUnhandledFn>(
+    context: Context,
+    unhandled_fn: NewUnhandledFn,
+) -> ConcreteResolvedTypeVisitor<
+    'resolver,
+    Context,
+    TypeId,
+    Output,
+    NewUnhandledFn,
+    impl FnOnce(Context) -> Output,
+    impl for<'a> FnOnce(Context, &'a mut dyn FieldIter<'resolver, TypeId>) -> Output,
+    impl for<'a> FnOnce(
+        Context,
+        &'a mut dyn VariantIter<'resolver, ConcreteFieldIter<'resolver, TypeId>>,
+    ) -> Output,
+    impl FnOnce(Context, &'resolver TypeId) -> Output,
+    impl FnOnce(Context, &'resolver TypeId, usize) -> Output,
+    impl for<'a> FnOnce(Context, &'a mut dyn ExactSizeIterator<Item = &'resolver TypeId>) -> Output,
+    impl FnOnce(Context, Primitive) -> Output,
+    impl FnOnce(Context, &'resolver TypeId) -> Output,
+    impl FnOnce(Context, BitsStoreFormat, BitsOrderFormat) -> Output,
+>
```

Comment:
This is insane, I didn't even know this kind of impl usage is possible in Rust. First I was confused of how fixing the types of all closures like this would work. But then I saw below that the type params of `Self` are mapped to new types in the builder functions below.

Reply:
James a.k.a the wizard ^^